src/main/java/org/opensearch/geospatial/ip2geo/cache/Ip2GeoCache.java

```diff
@@ -0,0 +1,120 @@
+/*
+ * Copyright OpenSearch Contributors
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+package org.opensearch.geospatial.ip2geo.cache;
+
+import java.io.IOException;
+import java.time.Instant;
+import java.util.Map;
+import java.util.concurrent.ConcurrentHashMap;
+
+import lombok.Getter;
+import lombok.extern.log4j.Log4j2;
+
+import org.opensearch.common.xcontent.LoggingDeprecationHandler;
+import org.opensearch.common.xcontent.XContentType;
+import org.opensearch.core.xcontent.NamedXContentRegistry;
+import org.opensearch.core.xcontent.XContentParser;
+import org.opensearch.geospatial.ip2geo.common.DatasourceFacade;
+import org.opensearch.geospatial.ip2geo.common.DatasourceState;
+import org.opensearch.geospatial.ip2geo.jobscheduler.Datasource;
+import org.opensearch.index.engine.Engine;
+import org.opensearch.index.shard.IndexingOperationListener;
+import org.opensearch.index.shard.ShardId;
+
+@Log4j2
+public class Ip2GeoCache implements IndexingOperationListener {
+    private final DatasourceFacade datasourceFacade;
+    private Map<String, DatasourceMetadata> data;
+
+    public Ip2GeoCache(final DatasourceFacade datasourceFacade) {
+        this.datasourceFacade = datasourceFacade;
+    }
+
+    public String getIndexName(final String datasourceName) {
+        return getData().getOrDefault(datasourceName, DatasourceMetadata.EMPTY_METADATA).getIndexName();
+    }
+
+    public boolean isExpired(final String datasourceName) {
+        return getData().getOrDefault(datasourceName, DatasourceMetadata.EMPTY_METADATA).getExpirationDate().isBefore(Instant.now());
+    }
+
+    public boolean has(final String datasourceName) {
+        return getData().containsKey(datasourceName);
+    }
+
+    public DatasourceState getState(final String datasourceName) {
+        return getData().getOrDefault(datasourceName, DatasourceMetadata.EMPTY_METADATA).getState();
+    }
+
+    private Map<String, DatasourceMetadata> getData() {
+        if (data != null) {
+            return data;
+        }
+        synchronized (this) {
+            if (data != null) {
+                return data;
+            }
+            Map<String, DatasourceMetadata> tempData = new ConcurrentHashMap<>();
+            datasourceFacade.getAllDatasources()
+                .stream()
+                .forEach(datasource -> tempData.put(datasource.getName(), new DatasourceMetadata(datasource)));
+            data = tempData;
+            return data;
+        }
+    }
+
+    private void put(final Datasource datasource) {
+        DatasourceMetadata metadata = new DatasourceMetadata(datasource);
+        getData().put(datasource.getName(), metadata);
+    }
+
+    private void remove(final String datasourceName) {
+        getData().remove(datasourceName);
+    }
```

Comment:
The put and remove methods directly access the data map without any synchronization, shall we ensure that only one thread can modify the data at one time?

Reply:
It is a concurrent hash map and it will handle concurrent behavior.